include/circt/Dialect/FIRRTL/FIRRTLAnnotations.h

```diff
@@ -63,6 +76,40 @@ class AnnotationSet {
   DictionaryAttr
   getArgumentAttrDict(ArrayRef<NamedAttribute> otherPortAttrs = {}) const;
 
+  /// Store this annotation set on an operation. If `removeOnEmpty` is set, the
+  /// annotation attribute is removed from the operation altogether. Returns
+  /// true if the operation was modified, false otherwise.
+  bool updateOperation(Operation *op, bool removeOnEmpty = true) const;
+
+  /// Store this annotation set in the attribute list of an operation. If
+  /// `removeOnEmpty` is set, the annotation attribute is removed from the
+  /// operation altogether if this set is empty. Returns true if the attribute
+  /// list was modified, false otherwise.
+  bool updateAttrs(NamedAttrList &attrs, bool removeOnEmpty = true) const;
+
+  /// Store this annotation set in the attribute list of a port. If
+  /// `removeOnEmpty` is set, the annotation attribute is removed from the port
+  /// altogether if this set is empty. Returns true if the attribute list was
+  /// modified, false otherwise.
+  bool updatePortAttrs(NamedAttrList &attrs, bool removeOnEmpty = true) const;
+
+  /// Update the attribute dictionary of an operation to contain this annotation
+  /// set. If `removeOnEmpty` is set, the annotation attribute is removed from
+  /// the operation altogether if this set is empty. Returns the updated
+  /// dictionary.
+  DictionaryAttr updateDictionaryAttr(DictionaryAttr attrs,
+                                      bool removeOnEmpty = true) const;
```

Comment:
I'd love to get rid of `removeOnEmpty`.  Is there any way to make this just do the right thing?  Why is this needed?

Reply:
I'll drop that. Pretty sure we want the `removeOnEmpty = true` behaviour in all cases anyway (to not leave empty `annotations = []` behind).